book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,220 @@
+}
+```
+
+We know by protocol (2nd table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus)) that V5 transactions will support sapling data however we also know by protocol that spends ([Spend Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#spendencodingandconsensus), See †) and outputs ([Output Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#outputencodingandconsensus), See †) fields change from V4 to V5.
+
+Here we have the proposed changes for V4 transactions:
+
+```
+V4 {
+    ...
+    sapling_value_balance: Amount,
+    ...
+    sapling_shielded_data: Option<SaplingShieldedDataV4>,
+}
+```
+
+`ShieldedData` is currently defined and implemented in `src/transaction/shielded_data.rs`. We propose to change `ShieldedData` into `SaplingShieldedDataV4` and to add `SaplingShieldedDataV5` into the same file.
+
+The difference between V4 and V5 shielded data is in the spends and outputs. For this reason we need to add V4 and V5 spend and output versions to our code:
+
+```
+pub struct SaplingShieldedDataV4 {
+    /// Either a spend or output description.
+    ///
+    /// Storing this separately ensures that it is impossible to construct
+    /// an invalid `ShieldedData` with no spends or outputs.
+    /// ...
+    pub first: Either<SpendV4, OutputV4>,
+    pub rest_spends: Vec<SpendV4>,
+    pub rest_outputs: Vec<OutputV4>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+Proposed `SpendV4` and `OutputV4` is now described, this is the same as the current implementation but just changing the struct names. 
+Sapling spend code is located at `zebra-chain/src/sapling/spend.rs`:
+
+```
+pub struct SpendV4 {
+    pub cv: commitment::ValueCommitment,
+    pub anchor: tree::Root,
+    pub nullifier: note::Nullifier,
+    pub rk: redjubjub::VerificationKeyBytes<SpendAuth>,
+    pub zkproof: Groth16Proof,
+    pub spend_auth_sig: redjubjub::Signature<SpendAuth>,
+}
+```
+
+The output code is located at `zebra-chain/src/sapling/output.rs`:
+
+```
+pub struct OutputV4 {
+    pub cv: commitment::ValueCommitment,
+    pub cm_u: jubjub::Fq,
+    pub ephemeral_key: keys::EphemeralPublicKey,
+    pub enc_ciphertext: note::EncryptedNote,
+    pub out_ciphertext: note::WrappedNoteKey,
+    pub zkproof: Groth16Proof,
+}
+```
+
+Now lets see how the v5 transaction is specified in the protocol, this is the second table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus).
+
+We propose the following representation for transaction V5 in Zebra:
+
+```
+V5 {
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    sapling_value_balance: Amount,
+    sapling_anchor: tree::Root,
+    sapling_shielded_data: Option<SaplingShieldedDataV5>,
+    orchard_flags: OrchardFlags,
+    orchard_value_balance: Amount,
+    orchard_anchor: tree::Root,
```

Comment:
@dconnolly should we move `sapling_anchor` and `orchard_anchor` into `ShieldedData`?
Also what about `orchard_flags` or `value_balance`?

How do we decide what data to put in the transaction, and what to put in `ShieldedData`?

Reply:
Since there is a `saplingAnchor` and an `orchardAnchor` yeah I think they respectively should go into their `ShieldedData`, and the `orchard_flags` should go in. Not quite sure about `value_balance` yet, hm

Reply:
@oxarbitrage can you make these changes?